### PR TITLE
Add pattern support for Meshery-Traefik-Mesh

### LINKED ISSUE #106: Add Patterns Support
**Current Behavior**
No Pattern Support

**Desired Behavior**
Add Pattern Support

---

### Contributor Resources

Refer: https://github.com/meshery/meshery-osm/pull/103

The meshery.io website uses Jekyll and GitHub Pages. Site content is found under the [`master` branch](https://github.com/meshery/meshery.io/tree/master).
- See the [Contributing to Meshery.io Website](https://github.com/layer5io/meshery.io#contributing-to-the-mesheryio-website) section of the readme.md and other [contributing instructions](https://docs.meshery.io/project/contributing), too.
- See Meshery site designs in this [Figma designfile](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI?node-id=110%3A1). Fill-in the [Layer5 Community Member Form](https://layer5.io/newcomer) and join the [Layer5 Community Slack](http://slack.layer5.io) for access.

### PR DESCRIPTION
**Pull request by Abhishek-kumar09**
**Description**
Add pattern support for traefik-mesh.

This PR fixes #106 

**Notes for Reviewers**

**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by Abhishek-kumar09** [comment]
Getting this error from running container
```console
INFO[2021-08-28T05:17:26Z] Adaptor Listening at port: 10006              app=traefik-mesh-adaptor
INFO[2021-08-28T05:17:26Z] lstat /.meshery/templates/oam/workloads: no such file or directory  app=traefik-mesh-adaptor
INFO[2021-08-28T05:17:26Z] lstat /.meshery/templates/oam/traits: no such file or directory  app=traefik-mesh-adaptor

```

**Comment by Abhishek-kumar09** [comment]
Traits and Workloads for traefik
Screenshot from 2021-08-28 13-35-13
Screenshot from 2021-08-28 13-33-24

**Comment by Revolyssup** [comment]
LGTM

**Review by Revolyssup — APPROVED** [review]

**Review by sayantan1413 — APPROVED** [review]

File: traefik/install.go [hunk 1]
```diff
@@ -64,91 +41,14 @@ func (mesh *Mesh) applyHelmChart(del bool, version, namespace string) error {
...
-		return ErrConvertingAppVersionToChartVersion(err)
-	}
-
-	err = kClient.ApplyHelmChart(mesherykube.ApplyHelmChartConfig{
+	return kClient.ApplyHelmChart(mesherykube.ApplyHelmChartConfig{
 		ChartLocation: mesherykube.HelmChartLocation{
 			Repository: repo,
 			Chart:      chart,
-			Version:    chartVersion,
+			Version:    version,
```
Comment (utkarsh-pro) [inline traefik/install.go:48]:
```suggestion
			AppVersion:    version,
```

File: traefik/sample-apps.go [hunk 2]
```diff
@@ -38,3 +42,45 @@ func (mesh *Mesh) applyManifest(contents []byte, isDel bool, namespace string) e
 
 	return nil
 }
+
+// sidecarInjection enables/disables sidecar injection on a namespace
+func (mesh *Mesh) sidecarInjection(namespace string, del bool) error {
```
Comment (utkarsh-pro) [inline traefik/sample-apps.go:47]: Traefik mesh doesn't have side car proxies instead relies on configuring CoreDNS to offer custom routing. Please remove this function.

File: traefik/oam.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+	return mergeMsgs(msgs), nil
+}
+
+// HandleApplicationConfiguration handles the processing of OAM application configuration
+func (mesh *Mesh) HandleApplicationConfiguration(config v1alpha1.Configuration, isDel bool) (string, error) {
+	var errs []error
+	var msgs []string
+	for _, comp := range config.Spec.Components {
+		for _, trait := range comp.Traits {
+			if trait.Name == "automaticSidecarInjection.Traefik" {
```
Comment (utkarsh-pro) [inline traefik/oam.go:58]: This trait should not exist. There is not sidecar injection in this mesh.

File: templates/oam/traits/automaticsidecarinjection.Traefik.meshery.layer5io.schema.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (utkarsh-pro) [inline templates/oam/traits/automaticsidecarinjection.Traefik.meshery.layer5io.schema.json:1]: Please remove this file.

File: templates/oam/traits/automaticsidecarinjection.Traefik_definition.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (utkarsh-pro) [inline templates/oam/traits/automaticsidecarinjection.Traefik_definition.json:1]: Please remove this file

File: main.go [hunk 6]
```diff
@@ -81,7 +84,7 @@ func main() {
 		os.Exit(1)
 	}
 
-	kubeconfigHandler, err := config.NewKubeconfigBuilder(configprovider.ViperKey)
+	kubeconfigHandler, err := config.NewKubeconfigBuilder(configprovider.InMemKey)
```
Comment (utkarsh-pro) [inline main.go:87]: Helm installations doesn't quite works with in-memory stores. Please use meshkit's viper config provider (refer: https://github.com/meshery/meshery-kuma/pull/193)

File: traefik/install.go (hunk 1 above)
Comment (Abhishek-kumar09) [reply traefik/install.go]: Done

File: traefik/sample-apps.go (hunk 2 above)
Comment (Abhishek-kumar09) [reply traefik/sample-apps.go]: Done

File: traefik/oam.go (hunk 3 above)
Comment (Abhishek-kumar09) [reply traefik/oam.go]: Removed

File: templates/oam/traits/automaticsidecarinjection.Traefik.meshery.layer5io.schema.json (hunk 4 above)
Comment (Abhishek-kumar09) [reply templates/oam/traits/automaticsidecarinjection.Traefik.meshery.layer5io.schema.json]: Removed.
Now trait doesn't have any file.
Should there be any other trait for Traefik?

File: templates/oam/traits/automaticsidecarinjection.Traefik_definition.json (hunk 5 above)
Comment (Abhishek-kumar09) [reply templates/oam/traits/automaticsidecarinjection.Traefik_definition.json]: removed

File: main.go (hunk 6 above)
Comment (Abhishek-kumar09) [reply main.go]: Done

File: templates/oam/traits/automaticsidecarinjection.Traefik.meshery.layer5io.schema.json (hunk 4 above)
Comment (utkarsh-pro) [reply templates/oam/traits/automaticsidecarinjection.Traefik.meshery.layer5io.schema.json]: I don't think there are any traits for traefik just yet.

File: main.go (hunk 6 above)
Comment (utkarsh-pro) [reply main.go]: Thanks for restoring `ViperKey` however as this is still using `meshery-adapter-library`'s `config` handlers, we will have data race issues. Let's track this issue in another issue.

Comment (Abhishek-kumar09) [reply main.go]: Currently Viper key is not written/changed by any function. So is there still a chance for data race issue?

Comment (utkarsh-pro) [reply main.go]: You are right. This isn't where we will get data race issue. However, `ViperKey` implies that we are instructing adapter to use `viper` as its config provider. This seems and works fine but starts panicking (due to data race) when concurrent actions are performed on the adapter. Adapter, in order to perform these actions uses its config provider to R/W some data, by default `viper` is **not** thread safe, hence it panics. Meshkit's viper wrapper deals with this thread unsafe nature.

**Review by utkarsh-pro — APPROVED** [review]
LGTM. Thank you @Abhishek-kumar09

**Pull request merged**